### PR TITLE
nimpackages / buildNimPackage

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

We really need a `buildNimPackage` function.

Related #67878

###### Things done

- [x] Write a section for the "Languages and frameworks" chapter of the Nixpkgs manual

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/nim-packages/build-nim-package/default.nix
Comment (erdnaxe) [inline pkgs/development/nim-packages/build-nim-package/default.nix:18]: There is two consecutive comma. This is making nixpkgs-review fail.

**Comment by erdnaxe** [comment]
Result of `nixpkgs-review pr 136489` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>38 packages built:</summary>
  <ul>
    <li>hottext</li>
    <li>mosdepth</li>
    <li>nimPackages.astpatternmatching</li>
    <li>nimPackages.bumpy</li>
    <li>nimPackages.chroma</li>
    <li>nimPackages.docopt</li>
    <li>nimPackages.flatty</li>
    <li>nimPackages.frosty</li>
    <li>nimPackages.hts-nim</li>
    <li>nimPackages.jester</li>
    <li>nimPackages.jsonschema</li>
    <li>nimPackages.karax</li>
    <li>nimPackages.lscolors</li>
    <li>nimPackages.markdown</li>
    <li>nimPackages.nim</li>
    <li>nimPackages.nimbox</li>
    <li>nimPackages.nimcrypto</li>
    <li>nimPackages.nimsimd</li>
    <li>nimPackages.noise</li>
    <li>nimPackages.packedjson</li>
    <li>nimPackages.pixie</li>
    <li>nimPackages.redis</li>
    <li>nimPackages.redpool</li>
    <li>nimPackages.regex</li>
    <li>nimPackages.sass</li>
    <li>nimPackages.sdl2</li>
    <li>nimPackages.segmentation</li>
    <li>nimPackages.supersnappy</li>
    <li>nimPackages.typography</li>
    <li>nimPackages.unicodedb</li>
    <li>nimPackages.unicodeplus</li>
    <li>nimPackages.vmath</li>
    <li>nimPackages.zippy</li>
    <li>nimlsp</li>
    <li>nimmm</li>
    <li>nitter</li>
    <li>nrpl</li>
    <li>swaycwd</li>
  </ul>
</details>

File: pkgs/development/nim-packages/build-nim-package/builder.nim
Comment (erdnaxe) [inline pkgs/development/nim-packages/build-nim-package/builder.nim:8]: Maybe we could precise that Nimble is BSD-licensed (I believe)?

File: pkgs/development/nim-packages/fetch-nimble/builder.sh
Comment (erdnaxe) [inline pkgs/development/nim-packages/fetch-nimble/builder.sh:4]: Could you add a comment about why `--noSSLCheck` is required?

File: pkgs/development/nim-packages/frosty/default.nix
Comment (erdnaxe) [inline pkgs/development/nim-packages/frosty/default.nix:3]: When I contributed Nitter to nixpkgs, I was not aware of the existence of fetch from Nimble. I believe it would be cleaner to fetch everything from Nimble in this case to ease the upgrade process of Nitter. I can do another pull request after this one to propose this change.

**Review by erdnaxe — CHANGES_REQUESTED** [review]
LGTM and I confirm that Nitter still works. I did not test other Nim application.

Maybe you could rebase the two fixup commits to get a cleaner history in this pull request?

(My review is weak because I am still new to nixpkgs)

File: pkgs/development/nim-packages/frosty/default.nix
Comment (ehmry) [reply pkgs/development/nim-packages/frosty/default.nix:3]: I would leave `fetchFromGitHub`where it already is. `fetchNimble` is a hack, it fetches every dependency of the requested package and then throws them away. It should yield the same hash as `fetchFromGitHub`, so the two are interchangeable.

**Comment by ehmry** [comment]
@erdnaxe thanks for the review.

File: pkgs/development/nim-packages/build-nim-package/default.nix
Comment (symphorien) [inline pkgs/development/nim-packages/build-nim-package/default.nix:22]: maybe you can set strictDeps to true by default like buildPythonPackage and buildRustPackage do by default (at least)

File: pkgs/development/nim-packages/packedjson/default.nix
Comment (symphorien) [inline pkgs/development/nim-packages/packedjson/default.nix:8]: those fetcher only packages have no maintainer nor license

File: pkgs/top-level/all-packages.nix
Comment (symphorien) [inline pkgs/top-level/all-packages.nix:12058]: maybe nimPackages should be a scope, which would allow to override nim, nimble or nim_builder.

Comment (symphorien) [reply pkgs/top-level/all-packages.nix:12058]: ah sorry it is one.

File: pkgs/development/nim-packages/build-nim-package/default.nix
Comment (symphorien) [inline pkgs/development/nim-packages/build-nim-package/default.nix:4]: do you think it is worth exposing it in the nimPackages scope to make it overrideable?

File: pkgs/applications/science/biology/mosdepth/default.nix
Comment (jbedo) [inline pkgs/applications/science/biology/mosdepth/default.nix:18]: I'm happy to continue maintaining these dependencies.

**Review by jbedo — APPROVED** [review]
LGTM, these are great improvements and tested mosdepth is functional (did not check other packages. I'm happy to continue maintaining hts-nim and docopt if desired.

File: pkgs/development/nim-packages/build-nim-package/default.nix
Comment (ehmry) [reply pkgs/development/nim-packages/build-nim-package/default.nix:22]: Good idea.

File: pkgs/development/nim-packages/packedjson/default.nix
Comment (ehmry) [reply pkgs/development/nim-packages/packedjson/default.nix:8]: Yes, but for now not a regression as these dependencies were completely hidden.

File: pkgs/development/nim-packages/build-nim-package/default.nix
Comment (ehmry) [reply pkgs/development/nim-packages/build-nim-package/default.nix:4]: Yes, it will probably need some adjustsments in the future.

Comment (symphorien) [reply pkgs/development/nim-packages/build-nim-package/default.nix]: fair enough

File: pkgs/development/nim-packages/packedjson/default.nix
Comment (symphorien) [reply pkgs/development/nim-packages/packedjson/default.nix:8]: fair enough

**Review by symphorien — APPROVED** [review]

**Review by joachimschmidt557 — APPROVED** [review]
LGTM! If desired, I can also keep maintaining `nimbox`, `lscolors` and `noise` in future.

File: .github/labeler.yml
Comment (SuperSandro2000) [inline .github/labeler.yml:79]: Please keep in mind that the github action will remove the nim label from all other paths.

File: doc/languages-frameworks/nim.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/nim.section.md:26]: This should use fetchFromSourceHut.

File: doc/languages-frameworks/nim.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/nim.section.md:56]:
```suggestion
[pkgs/top-level/nim-packages.nix](../../pkgs/top-level/nim-packages.nix) and implemented at
[pkgs/development/nim-packages](../../pkgs/development/nim-packages).
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:200]:
```suggestion
        # Remove configurations that clash with ours
        ./nim.cfg.patch
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:219]:
```suggestion
          cat >> config/config.nims << EOF
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:232]:
```suggestion
          EOF
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:235]:
```suggestion
          cat > config/nim.cfg << EOF
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:239]:
```suggestion
          EOF
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:244]:
```suggestion
          cat >> config/nim.cfg << EOF
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:254]:
```suggestion
          EOF
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:274]:
```suggestion
        # Used by nim-gdb
        "--prefix PATH : ${lib.makeBinPath [ buildPackages.gdb ]}:${placeholder "out"}/bin"

        # These libraries may be referred to by the standard library.
        # This is broken for cross-compilation because the package
        # set will be shifted back by nativeBuildInputs.
        "--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ openssl pcre ]}"

        # Use the custom configuration
        "--set NIM_CONFIG_PATH ${placeholder "out"}/etc/nim"

        # Fortify hardening appends -O2 to gcc flags which is unwanted for unoptimized nim builds.
        ''--set NIX_HARDENING_ENABLE "''${NIX_HARDENING_ENABLE/fortify}"''
```
gdb should probably be specific to that package

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:285]:
```suggestion
          local binname=$(basename $binpath)
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:319]:
```suggestion
        description = nim'.meta.description + " (${targetPlatform.config} wrapper)";
```

File: pkgs/development/compilers/nim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/nim/default.nix:327]:
```suggestion
  };
}
```

File: pkgs/development/nim-packages/build-nim-package/default.nix
Comment (SuperSandro2000) [inline pkgs/development/nim-packages/build-nim-package/default.nix:13]:
```suggestion
```

File: pkgs/development/nim-packages/fetch-nimble/builder.sh
Comment (SuperSandro2000) [inline pkgs/development/nim-packages/fetch-nimble/builder.sh:6]:
```suggestion
# TODO: bring in the certificates for Nimble to verify the fetch of
# the package list.
nimble --accept --noSSLCheck develop "${pkgname}@${version}"
```

File: pkgs/development/nim-packages/fetch-nimble/builder.sh
Comment (SuperSandro2000) [inline pkgs/development/nim-packages/fetch-nimble/builder.sh:2]:
```suggestion
export HOME=$(mktemp -d)
```

File: pkgs/development/nim-packages/fetch-nimble/default.nix
Comment (SuperSandro2000) [inline pkgs/development/nim-packages/fetch-nimble/default.nix:12]: This should use the standard phases.

File: pkgs/development/nim-packages/markdown/default.nix
Comment (SuperSandro2000) [inline pkgs/development/nim-packages/markdown/default.nix:6]: Use long hashes to prevent 404's in the future.

File: pkgs/top-level/nim-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/nim-packages.nix:13]:
```suggestion
    buildNimPackage = callPackage ../development/nim-packages/build-nim-package { };
    fetchNimble = callPackage ../development/nim-packages/fetch-nimble { };

    astpatternmatching = callPackage ../development/nim-packages/astpatternmatching { };
```

**Review by SuperSandro2000 — COMMENTED** [review]
Always use long revs. Short revs are not permitted.

version numbers should never start with a v in nixpkgs.

**Pull request merged**
